engines/savestate.cpp

```diff
@@ -35,24 +35,29 @@ SaveStateDescriptor::SaveStateDescriptor()
 }
 
 SaveStateDescriptor::SaveStateDescriptor(const MetaEngine *metaEngine, int slot, const Common::U32String &d)
-	: _slot(slot), _description(d), _isLocked(false), _playTimeMSecs(0) {
-	initSaveType(metaEngine);
+	: _slot(slot), _description(d), _isLocked(false), _playTimeMSecs(0), _saveType(kSaveTypeUndetermined) {
+	initSaveSlot(metaEngine);
 }
 
 SaveStateDescriptor::SaveStateDescriptor(const MetaEngine *metaEngine, int slot, const Common::String &d)
-	: _slot(slot), _description(Common::U32String(d)), _isLocked(false), _playTimeMSecs(0) {
-	initSaveType(metaEngine);
+	: _slot(slot), _description(Common::U32String(d)), _isLocked(false), _playTimeMSecs(0), _saveType(kSaveTypeUndetermined) {
+	initSaveSlot(metaEngine);
 }
 
-void SaveStateDescriptor::initSaveType(const MetaEngine *metaEngine) {
-	// Do not allow auto-save slot to be deleted or overwritten.
+void SaveStateDescriptor::initSaveSlot(const MetaEngine *metaEngine) {
 	if (!metaEngine && g_engine)
 		metaEngine = g_engine->getMetaEngine();
-	const bool autosave =
-			metaEngine && ConfMan.getInt("autosave_period") && _slot == metaEngine->getAutosaveSlot();
-	_isWriteProtected = autosave;
-	_saveType = autosave ? kSaveTypeAutosave : kSaveTypeRegular;
-	_isDeletable = !autosave;
+	int autosaveSlot = metaEngine ? metaEngine->getAutosaveSlot() : -1;
+	
+	if (autosaveSlot >= 0 && _slot == autosaveSlot) {
+		const bool autosaveEnabled = ConfMan.getInt("autosave_period");
+		// If autosaving enabled, do not allow autosave slot to be deleted or overwritten
+		_isWriteProtected = autosaveEnabled;
```

Comment:
Why are we changing this slot behaviour when auutosaves are not enabled?

Reply:
Since you're putting this feature in order, why not fix it?

Reply:
IIRC I kept the old behavior, only streamlined it, and made it consistent.

I don't have a strong opinion about this issue anyway.

Reply:
> However, why not take this opportunity to review the current approach for handling the autosave slot, by enforcing it only when absolutely necessary?

The problem is that the saves rename is a very risky operation and with a potentially confusing GUI message. Imagine when you accidentally turn the autosave off, and then you suddenly see the message about renames.

There is no significant advantage for the save slot 0, since we often provide hundreds or even thousands of save slots. Thus, it is best to have it designated for autosave.

As a result, the rename option will pop up only for a tiny fraction of users who happened to have non-autosave slots from many years ago.

As you went this far in improving the feature, may I ask you to straighten this behaviour so we finally merge this PR?

Reply:
>> The problem is that the saves rename is a very risky operation and with a potentially confusing GUI message. Imagine when you accidentally turn the autosave off, and then you suddenly see the message about renames.

> This isn't possible. The GUI message is only activated when autosaving is on (it's also why I prefer any user  access to be limited to when autosaving is on, so that issues like this can be handled if and when they happen).

You misunderstood me.
1. Somebody turns off autosaves. That makes slot 0 writeable
2. They create a save to saveslot 0
3. They turn on autosaves. Now they have to deal with the save renames

Save renames are part of the functionality that "salvages" a save which is in the way of the autosave, e.g. sits in that relevant slot but is not an autosave.

So yes, please revert this functionality.

As for you, as a dev, wanting to debug autosaves, well, you can then disable the code during your development, or rename the file: you're a dev, and you know what you're doing.

PS. I do not review PRs every day or even every week. Please bear with me